File: 141.环形链表.cpp
/*
 * @lc app=leetcode.cn id=141 lang=cpp
 *
 * [141] 环形链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        auto t1 = head;
        auto t2 = head;
        while (t1 != nullptr && t2 != nullptr) {
            t1 = t1->next;
            t2 = t2->next;
            if (t2 != nullptr) {
                t2 = t2->next;
            } else {
                return false;
            }
            if (t1 == t2) {
                return true;
            }
        }
        return false;
    }
};
// @lc code=end

File: 141.环形链表.py
#
# @lc app=leetcode.cn id=141 lang=python
#
# [141] 环形链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        l1, l2 = head, head
        while l1 and l2:
            l1 = l1.next
            l2 = l2.next
            if l2:
                l2 = l2.next
            else:
                return False
            if l1 == l2:
                return True
        return False        
# @lc code=end

